### DISCUSSION TITLE
Ordering of Subapps in a MultiApp System

### DISCUSSION THREAD

**Discussion by freiler** [idaholab/moose#16141] [Q&A General] [answered]
Hello all,

I want to solve a problem of 3 equations in a decoupled manner using a Multiapp. 

Subapp1 solves Equation 1. It needs u1 and provides u2.
Subapp2 solves Equation 2. It needs u2 and provides u3.
Subapp3 solves Equation 3. It needs u2 and u3, and provides u1. 

Then, I would like to have a Master that calls the three subapp solves and dictates the ordering of the solves. It is important that the 3 equations are sequential. 

Could someone please lead me to the way this can be done? Or maybe guide me to a good example that follows the same idea?
Thank you in advance,

Ramiro

**Comment by YaqiWang** [comment]
You can have a main app set equation 2 as the subapp, then add equation 1 as a subapp of the equation 2 subapp on timestep_begin, add equation 3 as a subapp of the equation 2 subapp on timestep_end. Eq. 1 and Eq. 3 input are on the same level of app tree, i.e. both subapps of the first level subapp of Eq. 2.

**Comment by zachmprince** [comment, marked as answer]
Yaqi's comment might be a little confusing, so here is a visualization:
```
              u1               u3
[ SubApp1 ]<------[ MainApp ]----->[ SubApp2 ]
[ (Eq.1)  ]------>[ (Eq.2)  ]<-----[ (Eq.3)  ]
              u2               u1
```
For the ordering, set SubApp1 `execute_on = timestep_begin` and SubApp2 `execute_on = timestep_end`

**Reply by ragusa** [reply]
Quick follow-up question: What if one has 4 equations? The sequence of solve to preserve is u1-->u2-->u3-->u4. Thanks!

**Reply by YaqiWang** [reply]
Using Actions might solve this problem. Actions can enforce the ordering of adding subapps, which will be used by MOOSE to execute the subapps. We can create an action that takes parameter like `inputs = 'input1 input2 input3 input4'` and `execute_on` to add the subapps sequentially. MOOSE currently does not have the ordering capability of multiapps on the same level and the same execute_on flag, i.e. MultiApp does not accept a parameter like 'input' in mesh generators.

**Reply by YaqiWang** [reply]
Oh, I forgot one thing: The data transfer happens only once both at the beginning and the end of executing multiapps at an execute_on flag. This means if we have 4 subapps on the same execute_on flag, the data outputted by one subapp will not immediately available to the next subapp until next Picard iteration. So basically we are doing Jacobi style sweeping through subapps at the same execute_on flag. So even we enforce the ordering, it does not matter anyway. Sorry about my previous reply.

**Reply by zachmprince** [reply]
I don't think actions would help in this case because all transfers with a certain execute flag happen at once, either before or after all the subapps have been executed. So the second subapp in the ordering would not receive information from first subapp's solve until until after all subapps have been executed.

**Reply by YaqiWang** [reply]
Right. Someone wanted the capability of doing data transfer among sibling subapps, i.e. let subapp itself initiate the transfer after it gets executed. That along with action can solve this issue. But that subapp level transfer is something @friedmud dislike alot and could complicate the design and usage significantly.

**Reply by zachmprince** [reply]
For the case of `u1->u2->u3->u4->u1` I would use a tree structure like this:
```
          u3                   u2
    ------------>[ SubApp1 ]------>[ SubSubApp1 ]
    |            [ Eq. (2) ]<------[   (Eq.1)   ]
    |                |         u1
    |          u2    |
[ MainApp ]<----------
[ (Eq. 3) ]<----------
     |         u4    |
     |               |
     ----------->[ SubApp 2]
          u3     [ (Eq.4)  ]
```

**Reply by freiler** [reply]
This makes sense. Thank you very much for your answers!

**Reply by freiler** [reply]
Following @zachmprince comment above:

> Yaqi's comment might be a little confusing, so here is a visualization:
> 
> ```
>               u1               u3
> [ SubApp1 ]<------[ MainApp ]----->[ SubApp2 ]
> [ (Eq.1)  ]------>[ (Eq.2)  ]<-----[ (Eq.3)  ]
>               u2               u1
> ```
> 
> For the ordering, set SubApp1 `execute_on = timestep_begin` and SubApp2 `execute_on = timestep_end`

I would like to transfer u1 in the scheme above to SubApp1 before the timestep ends, so that u1 is available for the upcoming timestep to properly solve for u2.

Is there any way to do this?

Thank you!

**Reply by freiler** [reply]
The transfer in order should happen in the following way: 

u2 From SubApp1 to MainApp
u3 & u2 From MainApp to SubApp2
u1 From SubApp2 to SubApp1 (somehow)

Then a new iteration may start.

**Reply by zachmprince** [reply]
As long as you don't set the `execute_on` flag in the transfer, the order of operation should be:

1. Transfer to_multiapp
1. Solve multiapp
1. Transfer from_multiapp

So if your input looks like this in the mainapp:
```
[MultiApps]
  [SubApp1]
    type = FullSolveMultiApp
    input = sub1.i
    execute_on = timestep_begin
  []
  [SubApp2]
    type = FullSolveMultiApp
    input = sub2.i
    execute_on = timestep_end
  []
[]

[Transfers]
  [u1_to_sub1]
    type = MultiAppCopyTransfer
    multiapp = SubApp1
    direction = to_multiapp
    to_variable = u1
    from_variable = u1
  []
  [u2_from_sub1]
    type = MultiAppCopyTransfer
    multiapp = SubApp1
    direction = from_multiapp
    to_variable = u2
    from_variable = u2
  []
  [u2_u3_to_sub2]
    type = MultiAppCopyTransfer
    multiapp = SubApp2
    direction = to_multiapp
    to_variable = 'u2 u3'
    from_variable = 'u2 u3'
  []
  [u1_from_sub2]
    type = MultiAppCopyTransfer
    multiapp = SubApp2
    direction = from_multiapp
    to_variable = u1
    from_variable = u1
  []
[]
```

The order of operations will be:

1. Transfer u1 to SubApp1 from MainApp
1. Solve SubApp1
1. Transfer u2 to MainApp from SubApp1
1. Solve MainApp
1. Transfer u2 and u3 to SubApp2 from MainApp
1. Solve SubApp2
1. Transfer u1 to MainApp from SubApp2

Repeated until convergence

**Reply by freiler** [reply]
@zachmprince thank you so much! I think I got it now.

**Comment by freiler** [comment]
@YaqiWang @zachmprince Thank you both for your answers Yaqi and Zach. This solves my current problem.

A follow-up question on this would be if there is another way to order a sequence of subapps than using the "execute_on" option. My question aims at what would happen for example if we needed to solve multiple uncoupled equations (at least more than 3). The 'execute_on' option for multiple decoupled equations seems limited for a hypothetical case of this kind. 

Thanks again,

Ramiro

**Comment by zachmprince** [comment]
`execute_on` is the only way to order the apps from the main app. But you can use a nested multiapp structure to facilitate more control on the ordering.  Something like in this image: [https://mooseframework.inl.gov/workshop/?_ga=2.38902757.731230092.1603987117-130448754.1603987117#/51/3](https://mooseframework.inl.gov/workshop/?_ga=2.38902757.731230092.1603987117-130448754.1603987117#/51/3)

**Comment by ragusa** [comment]
Thanks. To avoid repeating this until convergence, I suppose we can limit the Picard iterations. This should be done at the master input, I think. In which block and what’s that syntax? Thanks!

—jean ragusa


http://multiphysics.engr.tamu.edu
http://nustem.engr.tamu.edu
-.-.-
iPhone, iTypo, iApologize.

On Nov 13, 2020, at 12:50 PM, Zachary Prince <notifications@github.com> wrote:

﻿

As long as you don't set the execute_on flag in the transfer, the order of operation should be:

  1.  Transfer to_multiapp
  2.  Solve multiapp
  3.  Transfer from_multiapp

So if your input looks like this in the mainapp:

[MultiApps]
  [SubApp1]
    type = FullSolveMultiApp
    input = sub1.i
    execute_on = timestep_begin
  []
  [SubApp2]
    type = FullSolveMultiApp
    input = sub2.i
    execute_on = timestep_end
  []
[]

[Transfers]
  [u1_to_sub1]
    type = MultiAppCopyTransfer
    multiapp = SubApp1
    direction = to_multiapp
    to_variable = u1
    from_variable = u1
  []
  [u2_from_sub1]
    type = MultiAppCopyTransfer
    multiapp = SubApp1
    direction = from_multiapp
    to_variable = u2
    from_variable = u2
  []
  [u2_u3_to_sub2]
    type = MultiAppCopyTransfer
    multiapp = SubApp2
    direction = to_multiapp
    to_variable = 'u2 u3'
    from_variable = 'u2 u3'
  []
  [u1_from_sub2]
    type = MultiAppCopyTransfer
    multiapp = SubApp2
    direction = from_multiapp
    to_variable = u1
    from_variable = u1
  []
[]


The order of operations will be:

  1.  Transfer u1 to SubApp1 from MainApp
  2.  Solve SubApp1
  3.  Transfer u2 to MainApp from SubApp1
  4.  Solve MainApp
  5.  Transfer u2 and u3 to SubApp2 from MainApp
  6.  Solve SubApp2
  7.  Transfer u1 to MainApp from SubApp2

Repeated until convergence

**Reply by zachmprince** [reply]
Here are all the options regarding picard iterations:
![image](https://user-images.githubusercontent.com/12819067/99112246-9a665d80-25aa-11eb-8534-32a8456bcff5.png)

You can find more details [here](https://mooseframework.inl.gov/source/executioners/Steady.html)

**Reply by zachmprince** [reply]
Oh sorry, this is in the `[Executioner]` block
